# Подход 1: градиентный бустинг "в лоб"

## Что указать в отчете
### В отчете по данному этапу вы должны ответить на следующие вопросы:

### Какие признаки имеют пропуски среди своих значений? Что могут означать пропуски в этих признаках (ответьте на этот вопрос для двух любых признаков)?
#### 'first_blood_time', 'first_blood_team', 'first_blood_player1', 'first_blood_player2', 'radiant_bottle_time', 'radiant_courier_time', 'radiant_flying_courier_time', 'radiant_first_ward_time', 'dire_bottle_time','dire_courier_time','dire_flying_courier_time','dire_first_ward_time'.
Признак first_blood_time - игровое время первой крови. Первая кровь не была пролита в течение первых 5 минут матча. Возможно обе команды были настроены на пассивную игру и попытки нарастить преимущество в лейте.
Признак radiant_first_ward_time - время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля. Ни один из игроков команды dire не покупал "ward" в течение первых 5 минут матча. Игра без саппортов на низком рейтинге...

### Как называется столбец, содержащий целевую переменную?
radiant_win

### Как долго проводилась кросс-валидация для градиентного бустинга с 30 деревьями? Инструкцию по измерению времени можно найти ниже по тексту. Какое качество при этом получилось? Напомним, что в данном задании мы используем метрику качества AUC-ROC.
Затраченное время:  0:01:23, AUC = 0.69  

### Имеет ли смысл использовать больше 30 деревьев в градиентном бустинге? Что бы вы предложили делать, чтобы ускорить его обучение при увеличении количества деревьев?
С ростом числа деревьев, при сохранении остальных параметров увеличивается общее время расчета. Среднее значение AUC по кросс-валидации увеличивается при росте числа деревьев, что говорит нам о том, что использование большего числа деревьев увеличивает качество модели на обучающей выборке. Однако, с ростом качества на обучении увеличивается риск переобучения


## Загрузка обучающей и тестовой выборок

In [229]:
import pandas as pd

features = pd.read_csv('./features.csv', index_col='match_id')

features.head()

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   

          r1_kills  r1_deaths  r1_items          ...           \
match_id                                         ...            
0                0          0         7          ...            
1                0          1        12          ...            
2                0          0        12          ...            
3                0          0         5          ...            
4                1          0         8          ...            

          dire_boots_count  dire_ward_observer_count  dire_ward_sentry_count  \
match_id                                                                       
0                        4                         2                       2   
1                        4                         3                       1   
2                        4                         3                       1   
3                        4                         2                       0   
4                        3                         3                       0   

          dire_first_ward_time  duration  radiant_win  tower_status_radiant  \
match_id                                                                      
0                        -52.0      2874            1                  1796   
1                         -5.0      2463            1                  1974   
2                         13.0      2130            0                     0   
3                         27.0      1459            0                  1920   
4                        -16.0      2449            0                     4   

          tower_status_dire  barracks_status_radiant  barracks_status_dire  
match_id                                                                    
0                         0                       51                     0  
1                         0                       63                     1  
2                      1830                        0                    63  
3                      2047                       50                    63  
4                      1974                        3                    63  

[5 rows x 108 columns]

In [230]:
features_test = pd.read_csv('./features_test.csv', index_col='match_id')

features_test.head()

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
6         1430287923           0       93         4   1103     1089      8   
7         1430293357           1       20         2    556      570      1   
10        1430301774           1      112         2    751      808      1   
13        1430323933           1       27         3    708      903      1   
16        1430331112           1       39         4   1259      661      4   

          r1_kills  r1_deaths  r1_items          ...           \
match_id                                         ...            
6                0          1         9          ...            
7                0          0         9          ...            
10               0          0        13          ...            
13               1          1        11          ...            
16               0          0         9          ...            

          radiant_ward_sentry_count  radiant_first_ward_time  \
match_id                                                       
6                                 0                     12.0   
7                                 2                    -29.0   
10                                1                    -22.0   
13                                2                    -49.0   
16                                0                     36.0   

          dire_bottle_time  dire_courier_time  dire_flying_courier_time  \
match_id                                                                  
6                    247.0              -86.0                     272.0   
7                    168.0              -54.0                       NaN   
10                    46.0              -87.0                     186.0   
13                    30.0              -89.0                     210.0   
16                   180.0              -86.0                     180.0   

          dire_tpscroll_count  dire_boots_count  dire_ward_observer_count  \
match_id                                                                    
6                           3                 4                         2   
7                           3                 2                         2   
10                          1                 3                         3   
13                          3                 4                         2   
16                          1                 3                         2   

          dire_ward_sentry_count  dire_first_ward_time  
match_id                                                
6                              0                 118.0  
7                              1                  16.0  
10                             0                 -34.0  
13                             1                 -26.0  
16                             1                 -33.0  

[5 rows x 102 columns]

## Найдем признаки, отсутствующие в тестовой выборке.

In [231]:
itogs = [x for x in features.columns if x not in features_test.columns]

itogs

['duration',
 'radiant_win',
 'tower_status_radiant',
 'tower_status_dire',
 'barracks_status_radiant',
 'barracks_status_dire']

## Целевой признак - 'radiant_win'

In [232]:
y = features['radiant_win'].as_matrix()

y

array([1, 1, 0, ..., 0, 0, 1], dtype=int64)

## Оставим только те признаки, которые есть в тестовой выборке

In [233]:
features = features[[x for x in features.columns if x in features_test.columns]]

features.head()

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   

          r1_kills  r1_deaths  r1_items          ...           \
match_id                                         ...            
0                0          0         7          ...            
1                0          1        12          ...            
2                0          0        12          ...            
3                0          0         5          ...            
4                1          0         8          ...            

          radiant_ward_sentry_count  radiant_first_ward_time  \
match_id                                                       
0                                 0                     35.0   
1                                 0                    -20.0   
2                                 1                    -39.0   
3                                 0                    -30.0   
4                                 0                     46.0   

          dire_bottle_time  dire_courier_time  dire_flying_courier_time  \
match_id                                                                  
0                    103.0              -84.0                     221.0   
1                    149.0              -84.0                     195.0   
2                     45.0              -77.0                     221.0   
3                    124.0              -80.0                     184.0   
4                    182.0              -80.0                     225.0   

          dire_tpscroll_count  dire_boots_count  dire_ward_observer_count  \
match_id                                                                    
0                           3                 4                         2   
1                           5                 4                         3   
2                           3                 4                         3   
3                           0                 4                         2   
4                           6                 3                         3   

          dire_ward_sentry_count  dire_first_ward_time  
match_id                                                
0                              2                 -52.0  
1                              1                  -5.0  
2                              1                  13.0  
3                              0                  27.0  
4                              0                 -16.0  

[5 rows x 102 columns]

## Найдем все признаки в которых присутствуют пропуски

In [234]:
[k for k,v in dict(features.count() / features.shape[0]).items() if v < 1]

['first_blood_time',
 'first_blood_team',
 'first_blood_player1',
 'first_blood_player2',
 'radiant_bottle_time',
 'radiant_courier_time',
 'radiant_flying_courier_time',
 'radiant_first_ward_time',
 'dire_bottle_time',
 'dire_courier_time',
 'dire_flying_courier_time',
 'dire_first_ward_time']

## Возможная причина пропуска в данных:

### Признак first_blood_time - игровое время первой крови.

#### Первая кровь не была пролита в течение первых 5 минут матча. Возможно обе команды были настроены на пассивную игру и попытки нарастить преимущество в лейте.

### Признак radiant_first_ward_time - время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля

#### Ни один из игроков команды dire не покупал "ward" в течение первых 5 минут матча. Игра без саппортов на низком рейтинге...

## Заменяем пропуски в обучающей выборке на 0 и проверяем, что не осталось ни одного признака с пропусками.

In [235]:
features = features.fillna(0)

[k for k,v in dict(features.count() / features.shape[0]).items() if v < 1]

[]

## Обучение модели градиентного бустинга

In [243]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score
import numpy as np
from sklearn.model_selection import KFold, cross_val_score

kf = KFold(n_splits=5, shuffle=True, random_state=42)

X = features.as_matrix()

for n in [10,20,30,40,50,60, 70, 80, 90, 100]:
    
    start_time = datetime.datetime.now()
    
    gbc = GradientBoostingClassifier(n_estimators=n, verbose=False, random_state=42)
    
    model_scores = cross_val_score(gbc, X, y, cv=kf, scoring='roc_auc')
        
    print("Деревьев: ", n, " Среднее значение AUC: ", round(np.mean(model_scores),4), " Затраченное время: ", datetime.datetime.now() - start_time)

Деревьев:  10  Среднее значение AUC:  0.6649  Затраченное время:  0:00:30.226027
Деревьев:  20  Среднее значение AUC:  0.6825  Затраченное время:  0:00:57.912176
Деревьев:  30  Среднее значение AUC:  0.69  Затраченное время:  0:01:23.821708
Деревьев:  40  Среднее значение AUC:  0.694  Затраченное время:  0:02:18.935141
Деревьев:  50  Среднее значение AUC:  0.6975  Затраченное время:  0:03:47.435787
Деревьев:  60  Среднее значение AUC:  0.7001  Затраченное время:  0:03:43.914903
Деревьев:  70  Среднее значение AUC:  0.7021  Затраченное время:  0:03:07.210351
Деревьев:  80  Среднее значение AUC:  0.7038  Затраченное время:  0:03:31.174732
Деревьев:  90  Среднее значение AUC:  0.7052  Затраченное время:  0:03:55.950579
Деревьев:  100  Среднее значение AUC:  0.7062  Затраченное время:  0:04:36.885805


## С ростом числа деревьев, при сохранении остальных параметров увеличивается общее время расчета. Среднее значение AUC по кросс-валидации увеличивается при росте числа деревьев, что говорит нам о том, что использование большего числа деревьев увеличивает качество модели на обучающей выборке. Однако, с ростом качества на обучении увеличивается риск переобучения

# Подход 2: логистическая регрессия

## Что указать в отчете
### В отчете по данному этапу вы должны ответить на следующие вопросы:

### Какое качество получилось у логистической регрессии над всеми исходными признаками? Как оно соотносится с качеством градиентного бустинга? Чем вы можете объяснить эту разницу? Быстрее ли работает логистическая регрессия по сравнению с градиентным бустингом?
Лучшее значение AUC получилось 0.7166 при С = 0.01. На модели бустинга подобное значение качества не было достигнуто при количестве деревьев меньше 100. Логистическая регрессия значительно быстрее бустинга.

### Как влияет на качество логистической регрессии удаление категориальных признаков (укажите новое значение метрики качества)? Чем вы можете объяснить это изменение?
Их удаление практически не повлияло на итоговые результаты, что говорит о низком влиянии данных показателей на модель. Лучшее значение AUC получилось 0.7166 при С = 0.01.

### Сколько различных идентификаторов героев существует в данной игре?
112

### Какое получилось качество при добавлении "мешка слов" по героям? Улучшилось ли оно по сравнению с предыдущим вариантом? Чем вы можете это объяснить?
Применение "мешка слов" привело к улучшению результатов модели на кросс-валидации. Лучшее значение AUC получилось 0.7519   при С = 0.1, что может быть объяснено включением в модель информации о составе каждоый из команд, что определенно должно повлиять на результаты матчей

### Какое минимальное и максимальное значение прогноза на тестовой выборке получилось у лучшего из алгоритмов?
минимум = 0.0086, максимум = 0.9965


In [241]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

scaler = StandardScaler()

X = scaler.fit_transform(features.as_matrix())

for C in [10.0 ** i for i in range(-5,5)]:
    
    start_time = datetime.datetime.now()
    
    lr = LogisticRegression(C=C, random_state=42, penalty='l2')
    
    model_scores = cross_val_score(lr, X, y, cv=kf, scoring='roc_auc')
        
    print("C: ", C, " Среднее значение AUC: ", round(np.mean(model_scores),4), " Затраченное время: ", datetime.datetime.now() - start_time)

C:  1e-05  Среднее значение AUC:  0.6952  Затраченное время:  0:00:02.948323
C:  0.0001  Среднее значение AUC:  0.7114  Затраченное время:  0:00:04.840465
C:  0.001  Среднее значение AUC:  0.7164  Затраченное время:  0:00:08.840887
C:  0.01  Среднее значение AUC:  0.7166  Затраченное время:  0:00:11.191529
C:  0.1  Среднее значение AUC:  0.7165  Затраченное время:  0:00:13.025532
C:  1.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:13.012543
C:  10.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.617107
C:  100.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.392449
C:  1000.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.346518
C:  10000.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.547783


## Удалим категориальные переменные

In [246]:
X = features[[x for x in features.columns if x not in ['lobby_type', 'r1_hero', 'r2_hero', 
                                                       'r3_hero','r4_hero','r5_hero','d1_hero', 
                                                       'd2_hero', 'd3_hero','d4_hero','d5_hero']]]

X = scaler.fit_transform(X.as_matrix())

for C in [10.0 ** i for i in range(-5,5)]:
    
    start_time = datetime.datetime.now()
    
    lr = LogisticRegression(C=C, random_state=42, penalty='l2')
    
    model_scores = cross_val_score(lr, X, y, cv=kf, scoring='roc_auc')
        
    print("C: ", C, " Среднее значение AUC: ", round(np.mean(model_scores),4), " Затраченное время: ", datetime.datetime.now() - start_time)

C:  1e-05  Среднее значение AUC:  0.6951  Затраченное время:  0:00:02.706441
C:  0.0001  Среднее значение AUC:  0.7113  Затраченное время:  0:00:04.324435
C:  0.001  Среднее значение AUC:  0.7164  Затраченное время:  0:00:08.095685
C:  0.01  Среднее значение AUC:  0.7166  Затраченное время:  0:00:10.183512
C:  0.1  Среднее значение AUC:  0.7165  Затраченное время:  0:00:11.189716
C:  1.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.417006
C:  10.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.689146
C:  100.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.441751
C:  1000.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.257701
C:  10000.0  Среднее значение AUC:  0.7165  Затраченное время:  0:00:12.245718


## Добавим "мешок слов"

In [252]:
heroes = pd.read_csv('./data/dictionaries/heroes.csv')
print('Всего героев в игре:', len(heroes))

Всего героев в игре: 112


In [254]:
N = len(heroes)

X_pick = np.zeros((features.shape[0], N))

for i, match_id in enumerate(features.index):
    
    for p in range(5):
        
        X_pick[i, features.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick[i, features.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
        
X_pick

C:\Users\ccord517\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  if __name__ == '__main__':


array([[ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       ...,
       [ 0.,  0.,  0., ...,  0.,  0.,  1.],
       [ 0.,  0.,  0., ...,  0.,  0.,  0.],
       [ 0.,  0.,  0., ...,  0.,  0., -1.]])

In [272]:
X = pd.concat([pd.DataFrame(X, index=features.index), pd.DataFrame(X_pick, index=features.index)], axis=1).as_matrix()

X

array([[-2.54436416,  1.40080818,  1.52597175, ...,  0.        ,
         0.        ,  0.        ],
       [-2.54045236,  0.50131354, -0.08013929, ...,  0.        ,
         0.        ,  0.        ],
       [-2.53923104,  0.50131354,  0.1510701 , ...,  0.        ,
         0.        ,  0.        ],
       ...,
       [ 1.09874571,  0.50131354,  0.29226667, ...,  0.        ,
         0.        ,  1.        ],
       [ 1.09895204, -0.39818111, -0.17368203, ...,  0.        ,
         0.        ,  0.        ],
       [ 1.1026479 , -0.39818111, -0.31840851, ...,  0.        ,
         0.        , -1.        ]])

In [276]:
for C in [10.0 ** i for i in range(-5,5)]:
    
    start_time = datetime.datetime.now()
    
    lr = LogisticRegression(C=C, random_state=42, penalty='l2')
    
    model_scores = cross_val_score(lr, X, y, cv=kf, scoring='roc_auc')
        
    print("C: ", C, " Среднее значение AUC: ", round(np.mean(model_scores),4), " Затраченное время: ", datetime.datetime.now() - start_time)

C:  1e-05  Среднее значение AUC:  0.6992  Затраченное время:  0:00:03.793696
C:  0.0001  Среднее значение AUC:  0.7251  Затраченное время:  0:00:05.647936
C:  0.001  Среднее значение AUC:  0.7463  Затраченное время:  0:00:10.390910
C:  0.01  Среднее значение AUC:  0.7517  Затраченное время:  0:00:16.175406
C:  0.1  Среднее значение AUC:  0.7519  Затраченное время:  0:00:22.302058
C:  1.0  Среднее значение AUC:  0.7519  Затраченное время:  0:00:23.056440
C:  10.0  Среднее значение AUC:  0.7519  Затраченное время:  0:00:23.079859
C:  100.0  Среднее значение AUC:  0.7519  Затраченное время:  0:00:24.099362
C:  1000.0  Среднее значение AUC:  0.7519  Затраченное время:  0:00:23.945394
C:  10000.0  Среднее значение AUC:  0.7519  Затраченное время:  0:00:23.444619


## Предскажем вероятность победы radiant на тестовой выборке по лучшей модели

In [277]:
lr = LogisticRegression(C=0.1, random_state=42, penalty='l2')

lr.fit(X,y)

LogisticRegression(C=0.1, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=42, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [303]:
features_test = features_test.fillna(0)

X_test = features_test[[x for x in features_test.columns if x not in ['lobby_type', 'r1_hero', 'r2_hero', 
                                                       'r3_hero','r4_hero','r5_hero','d1_hero', 
                                                       'd2_hero', 'd3_hero','d4_hero','d5_hero']]]

X_test = scaler.fit_transform(X_test.as_matrix())

X_pick_test = np.zeros((features_test.shape[0], N))

for i, match_id in enumerate(features_test.index):
    
    for p in range(5):
        
        X_pick_test[i, features_test.ix[match_id, 'r%d_hero' % (p+1)]-1] = 1
        X_pick_test[i, features_test.ix[match_id, 'd%d_hero' % (p+1)]-1] = -1
        
X_test = pd.concat([pd.DataFrame(X_test, index=features_test.index), pd.DataFrame(X_pick_test, index=features_test.index)], axis=1).as_matrix()

X_test

C:\Users\ccord517\Anaconda3\lib\site-packages\ipykernel_launcher.py:15: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  from ipykernel import kernelapp as app


array([[-2.5148749 ,  0.47474566, -0.23774298, ...,  0.        ,
         0.        ,  0.        ],
       [-2.51389501, -1.36563036, -1.20791332, ...,  0.        ,
         0.        ,  0.        ],
       [-2.5123772 , -1.36563036, -0.86205735, ...,  0.        ,
         0.        ,  1.        ],
       ...,
       [ 1.08005567, -1.36563036, -1.17421453, ...,  0.        ,
         0.        ,  0.        ],
       [ 1.08387463,  0.47474566,  1.0765097 , ...,  0.        ,
         0.        , -1.        ],
       [ 1.08579655,  0.47474566, -0.0390975 , ...,  0.        ,
         0.        ,  1.        ]])

In [314]:
pred = lr.predict_proba(X_test)[:,1]

pd.concat([pd.DataFrame(X_test, index=features_test.index), pd.DataFrame({'radiant_win ':pred}, index=features_test.index)], axis=1)

0         1         2         3         4         5  \
match_id                                                               
6        -2.514875  0.474746 -0.237743 -0.137094 -0.370005 -0.532732   
7        -2.513895 -1.365630 -1.207913 -1.255660 -1.145090 -0.532732   
10       -2.512377 -1.365630 -0.862057 -0.742715 -1.145090 -0.532732   
13       -2.508381 -0.445442 -0.938323 -0.537967 -1.145090  0.968865   
16       -2.507087  0.474746  0.038942 -1.059534 -0.812911 -0.532732   
18       -2.506518  1.394934  1.472027  0.854314  0.405079 -0.532732   
19       -2.506387  0.474746  1.108435  0.720690  1.401616 -0.532732   
24       -2.503777  1.394934  1.337232  1.789686  0.847984  2.470463   
33       -2.500708 -1.365630 -1.277085 -1.406527 -1.255816 -0.532732   
37       -2.499155 -0.445442 -0.314009 -0.994877 -0.702184 -0.532732   
41       -2.498553  0.474746  0.927526  0.020239  0.515805 -0.532732   
42       -2.498346 -2.285818 -1.943966 -1.150054 -1.145090 -0.532732   
55       -2.495843 -1.365630 -1.078439 -1.283678 -1.255816 -0.532732   
60       -2.494946 -1.365630 -1.179535 -0.699610 -0.702184 -0.532732   
62       -2.494703 -0.445442  0.585217  1.444849  2.065974 -0.532732   
63       -2.494470  1.394934  1.762901  1.921155  1.180164 -0.532732   
64       -2.494230  0.474746  0.351099  0.043946  0.626532 -0.532732   
71       -2.493088  1.394934  1.083604  0.742242  1.401616 -0.532732   
72       -2.493079 -1.365630 -1.330293 -1.408682 -1.255816 -0.532732   
83       -2.490861 -1.365630 -1.686791 -1.281523 -1.255816 -0.532732   
85       -2.490726 -1.365630 -0.611977 -0.895736 -1.145090 -0.532732   
89       -2.490289  1.394934  1.255645  2.067711  1.512343  2.470463   
92       -2.489745 -0.445442 -0.390274  0.578444  0.958711 -0.532732   
100      -2.488734 -1.365630 -1.429616 -0.363393 -1.255816  0.968865   
102      -2.488332 -0.445442 -0.815943 -1.318162 -1.145090 -0.532732   
108      -2.487188 -1.365630 -1.388823 -0.887115 -1.145090 -0.532732   
111      -2.486977 -1.365630 -1.002173 -1.326783 -1.034363 -0.532732   
126      -2.484692 -1.365630 -0.510881 -0.555209 -1.034363  0.968865   
130      -2.484012 -1.365630 -1.371086 -1.216866 -1.034363 -0.532732   
140      -2.482653  0.474746  0.372382  1.397433  2.065974 -0.532732   
...            ...       ...       ...       ...       ...       ...   
114165    1.063467  0.474746  0.241135  0.589220  0.072900  0.968865   
114168    1.063591 -0.445442 -0.977343 -1.119881 -1.255816 -0.532732   
114171    1.063745  0.474746  1.053453 -0.143559  0.072900 -0.532732   
114183    1.064220 -1.365630 -1.067797 -0.984101 -0.812911 -0.532732   
114188    1.064413  0.474746 -0.083438  0.936213  1.512343 -0.532732   
114192    1.064532 -1.365630 -1.548448 -1.055224 -0.812911  0.968865   
114195    1.064825 -1.365630 -0.954286 -1.055224 -1.255816 -0.532732   
114202    1.065126  1.394934  1.546519  1.425451  1.069437  0.968865   
114209    1.065302 -0.445442 -0.276763 -0.548743 -0.923637 -0.532732   
114211    1.065349 -0.445442  0.014111  0.554737  0.958711 -0.532732   
114217    1.065654  1.394934  1.789505  2.444877  1.623069  0.968865   
114232    1.066426 -0.445442 -0.278536  0.429733  0.405079  0.968865   
114236    1.066619 -0.445442 -0.837227 -0.813837 -0.702184 -0.532732   
114238    1.066708  0.474746  0.375930  1.115098  1.180164  0.968865   
114242    1.066880 -0.445442 -0.244837  0.248693  0.072900 -0.532732   
114247    1.067154 -0.445442 -0.047966  0.617239  0.847984 -0.532732   
114285    1.068975  1.394934  0.863675  0.556892  1.180164 -0.532732   
114286    1.069032  0.474746  0.565707 -0.647884 -0.259279 -0.532732   
114314    1.070724  1.394934  0.913337  0.128000  0.294353 -0.532732   
114327    1.072114 -0.445442 -0.652770 -0.518570 -0.923637  0.968865   
114330    1.072400 -1.365630 -0.963154 -1.109104 -0.923637 -0.532732   
114337    1.073194 -1.365630 -0.840774 -0.186664 -1.034363  2.470463   
114348    1.074142  0.474746  0.698729  0.293953  0.84798

In [318]:
round(min(pred),4), round(max(pred),4)

(0.0086, 0.9965)